### ISSUE TITLE
DFL, Distribution Focal Loss

### ISSUE THREAD

**Issue by dylanjaybaxter** [ultralytics/ultralytics#6596]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I have been looking into the v8DetectionLoss module and I noticed something unexpected in the DFL implementation. Based on the linked paper and what I can find about DFL, I expect there to be some mechanism to address class distribution imbalance, but it seems like the implementation in BboxLoss is just taking a weighted average of the Cross Entropy Losses using floor and ceiling of the ground truth coordinates. 

Therefore, it's unclear to me where the class and easy-hard sample balancing is taking place, if it is at all. Maybe that effect is present in the softmax present in the pytorch Cross Entropy? If someone could point me to a resource or the part of the code that is responsible for those aspects of DFL, it would be greatly appreciated. 

Thank you!

**Comment by menggui1993** [comment]
DFL(distribution focal loss) is used in the bounding box regression. The general idea is to predict distributions of the box offsets instead of predicting the values directly. It has nothing to do with the object class classification.

I think what you are talking about is QFL(quality focal loss) which is introduced in the same paper as DFL. It is used for classification. As far as I'm concerned, YOLOv8 is not using QFL. They just use a plane Binary Cross Entropy for the classification head as you said.

You can easily replace it with something like QFL or VarifocalLoss by yourself in the code.
https://github.com/ultralytics/ultralytics/blob/fdcf0dd4fda06c4c67bde061fe98f9ddd77b2ea4/ultralytics/utils/loss.py#L196

**Comment by menggui1993** [comment]
I personally think the comment is not correct. As DFL has nothing to do with the classification and also in the code, it's just BCE used for classification.

In the code, I found they had an implementation of VarifocalLoss which is a loss used to deal with class imbalance. https://github.com/ultralytics/ultralytics/blob/fdcf0dd4fda06c4c67bde061fe98f9ddd77b2ea4/ultralytics/utils/loss.py#L15-L33
But it is not being used in the actual loss calculation. From this https://github.com/ultralytics/ultralytics/issues/1448#issuecomment-1471570296, it seems that they had experienced and found out the VarifocalLoss not showing any improvements. I guess that's why they just kept using BCE.

As for the comment, maybe we need the authors to confirm and clarify this.

**Comment by dylanjaybaxter** [comment]
Agreed. Thank you for helping to clear things up! I'll leave this issue open for a bit in the hopes that an author might notice it.

**Comment by glenn-jocher** [comment]
@dylanjaybaxter hi there,

Thank you for raising your concerns and allowing the opportunity to clarify. You are correct that DFL, or Distribution Focal Loss, is associated with bounding box regression and not with class imbalance or sample difficulty. In YOLOv8's implementation, DFL is indeed focused on the bounding box component rather than classification.

The comment you've referenced seems to be a misunderstanding. It conflates DFL with mechanisms designed to address class imbalance, such as Focal Loss or Quality Focal Loss (QFL). Focal Loss and its variants like QFL are geared towards solving the issue of class imbalance by modulating the loss based on the confidence of the classification, but this is separate from the technique utilized by DFL.

In YOLOv8, Binary Cross Entropy (BCE) is used for the classification loss, which does not incorporate the same focus on class imbalance as Focal Losses do. While other loss functions, such as VarifocalLoss, have been explored, they have not been implemented in YOLOv8's default configuration as they did not present a significant improvement over BCE in our experiments.

I hope this has helped resolve the confusion. The community and YOLOv8's continuous evolution is very much fueled by discussions like these, and I appreciate your engagement. If you have any other questions or need further explanations, please feel free to continue the discussion on our issue tracker. 😊

**Comment by dylanjaybaxter** [comment]
@glenn-jocher Hello,

Thank you so much for the clarification! That makes sense. I think I even remember reading that focal loss was used to no significant benefit as far back as yolov3. 

Also, I appreciate the welcoming words and look forward having more discussions like this going forward!

**Comment by glenn-jocher** [comment]
@dylanjaybaxter you're very welcome!

I'm glad I could help clear up any confusion. It's great to hear of your ongoing interest and engagement with the project. Discussions and insights from users like you are invaluable for driving the project forward and improving the tools we all rely on. Feel free to reach out anytime you have more questions or ideas to share.

Happy detecting, and here's to many more fruitful discussions in the future! 🚀😊

**Comment by YQ0Sun** [comment]
Hi, I would like to ask a question about the definition of regression loss function in YOLOv8, what kind of expressions are DFL Loss and CIoU Loss? Looking forward to your reply.

**Comment by glenn-jocher** [comment]
@YQ0Sun hi there! 👋

In YOLOv8, the regression loss function is composed of two main components: DFL (Distribution Focal Loss) and CIoU (Complete Intersection over Union) Loss.

- **DFL Loss** is used for bounding box regression. Instead of directly predicting the box coordinates, it predicts the distribution of possible box offsets, which helps in capturing the uncertainty in the box location.

- **CIoU Loss** measures the dissimilarity between the predicted bounding boxes and the ground truth. It takes into account the overlap area, the distance between the box centers, and the aspect ratio, providing a more comprehensive measure than traditional IoU.

Here's a simplified example of how CIoU loss might be calculated for two boxes `box1` and `box2`:

```python
import torch
from ultralytics.utils.general import bbox_ciou

box1 = torch.tensor([50, 50, 100, 100])  # x1, y1, x2, y2 format
box2 = torch.tensor([60, 60, 110, 110])  # x1, y1, x2, y2 format

ciou_loss = 1 - bbox_ciou(box1, box2)
print(f"CIoU Loss: {ciou_loss.item()}")
```

Hope this helps! Let me know if you have any more questions. 😊

**Comment by AgusKurniawan08** [comment]
Hello, I want to ask, on my machine learning graph it shows that val/dfl loss increases after 50 epochs. Is that a sign that my machine learning is bad?
and for information the results from the map are 93%
![Uploading Capture.PNG…]()

**Comment by glenn-jocher** [comment]
Hi there! 👋

Seeing an increase in your val/dfl loss after 50 epochs while having a mAP of 93% suggests that your model is performing quite well in terms of accuracy. 🎯 However, the increasing loss could indicate overfitting past those epochs. It means your model might be learning the training data too well, to the detriment of its performance on unseen data.

To mitigate this, you could consider strategies like early stopping (stop training when val/dfl loss starts increasing), regularization techniques, or introducing more variety into your training data.

Remember, a single metric doesn't tell the whole story, so your high mAP is definitely something to be proud of! 🌟

**Comment by AgusKurniawan08** [comment]
so it's ok if I claim my machine learning model is already good?

**Comment by glenn-jocher** [comment]
@AgusKurniawan08 absolutely! A high mAP of 93% is an excellent indicator of your model's performance. 😊 Just keep an eye on the validation loss for any signs of overfitting, but overall, it sounds like you're on the right track. Great job! 👍

**Comment by Mehreganbeiki** [comment]
Hi @glenn-jocher, Can you explain the loss functions that the YOLOv8 model uses when performing segmentation? Is it using the BCE as well as the DFL and CIoU? Can you further explain the DFL and CIoU in the segmentation model?

**Comment by daikankan** [comment]
For what kind of dataset can(should) we disable DFL(Distribution Focal Loss)? such as face detection task which bboxes are mainly squares @glenn-jocher

**Comment by glenn-jocher** [comment]
Hi @daikankan,

DFL is generally beneficial for most object detection tasks as it helps model bounding box uncertainty, even for square boxes like faces. However, if your dataset has extremely consistent box dimensions with minimal variance (e.g., tightly cropped fixed-aspect-ratio faces), you could consider disabling DFL by modifying the `loss.py` module. For most real-world cases where some box variation exists, we recommend keeping DFL enabled. You can find more details about loss functions in our [Loss Function documentation](https://www.ultralytics.com/glossary/loss-function).

**Comment by AvinashMehta2000** [comment]
I have fine-tuned the YOLO11 model for my real-world application, and all the metrics- accuracy and loss, are showing good results. I applied it to a large unseen test set and it performs very well, with an F1 score of over 92%. However, the issue I’ve noticed is a significant increase in the val/dfl loss (val distribution focal). How can I explain and report this in a clear and reasonable manner?

![Image](https://github.com/user-attachments/assets/ccac1b22-0d0b-4cd5-ba0d-a1218dcbd1fe)

**Comment by glenn-jocher** [comment]
The val/dfl loss increase while maintaining strong detection metrics (F1 @AvinashMehta2000%) suggests your model is achieving good practical performance despite potential overfitting in box regression confidence estimates. This divergence is common in real-world applications where: 

1. Final box predictions remain accurate (high mAP/F1) 
2. Model becomes less certain about exact box distributions during later epochs

You can report this as: *"While validation metrics indicate excellent detection performance (F1@92%), rising DFL loss suggests decreasing confidence in precise box offset distributions during later training stages - a known behavior that doesn't necessarily impact final detection accuracy when using weighted sum predictions."* 

For deployment-focused applications, the strong F1 score should be your primary metric. If concerned, consider early stopping at the DFL loss minimum or adjusting the `dfl` loss weight in your hyperparameters.

**Comment by Acui001** [comment]
> The val/dfl loss increase while maintaining strong detection metrics (F1 [@AvinashMehta2000](https://github.com/AvinashMehta2000)%) suggests your model is achieving good practical performance despite potential overfitting in box regression confidence estimates. This divergence is common in real-world applications where:val/dfl 损失增加，同时保持强大的检测指标 （F1 %） 表明，尽管盒回归置信度估计中可能存在过拟合，但您的模型仍实现了良好的实际性能。这种差异在实际应用中很常见，其中：
> 
> 1. Final box predictions remain accurate (high mAP/F1)最终框预测保持准确（高 mAP/F1）
> 2. Model becomes less certain about exact box distributions during later epochs模型在以后的 epoch 中对精确的盒分布变得不太确定
> 
> You can report this as: _"While validation metrics indicate excellent detection performance (F1@92%), rising DFL loss suggests decreasing confidence in precise box offset distributions during later training stages - a known behavior that doesn't necessarily impact final detection accuracy when using weighted sum predictions."_您可以将其报告为 _：“虽然验证指标表明出色的检测性能 （F1@92%），但不断上升的 DFL 损失表明在后期训练阶段对精确箱形偏移分布的置信度降低 - 这种已知行为在使用加权和预测时不一定会影响最终检测准确性。_
> 
> For deployment-focused applications, the strong F1 score should be your primary metric. If concerned, consider early stopping at the DFL loss minimum or adjusting the `dfl` loss weight in your hyperparameters.对于以部署为中心的应用程序，较高的 F1 分数应该是您的主要指标。如果担心，请考虑在 DFL 损失最小值处提前停止，或在超参数中调整 `DFL` 损失权重。

I want to know the DFL and QFL from the GFL, why code only use the DFL ?dose the QFL haven`t work ,same as VFL https://github.com/ultralytics/ultralytics/issues/1448#issuecomment-1471570296 not work?

**Comment by glenn-jocher** [comment]
@Acui001 you're correct that only DFL from the GFL paper is implemented in YOLOv8. QFL (Quality Focal Loss) is designed for classification tasks to address class imbalance, but similar to VFL (Varifocal Loss), our experiments showed QFL didn't provide significant improvements over standard BCE for classification in YOLOv8's architecture.

The main reasons are:
1. DFL works well for bbox regression uncertainty modeling 
2. QFL/VFL for classification didn't show meaningful gains in our testing
3. BCE provides sufficient performance with simpler implementation

You can find the unused [VarifocalLoss implementation](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.VarifocalLoss) in our codebase if you'd like to experiment with focal-based classification losses for your specific use case.
